src/execution/join_hashtable.cpp

```diff
@@ -275,32 +270,156 @@ void JoinHashTable::Finalize() {
 	auto hash_data = FlatVector::GetData<hash_t>(hashes);
 	data_ptr_t key_locations[STANDARD_VECTOR_SIZE];
 	// now construct the actual hash table; scan the nodes
-	// as we can the nodes we pin all the blocks of the HT and keep them pinned until the HT is destroyed
+	// as we scan the nodes we pin all the blocks of the HT and keep them pinned until the HT is destroyed
 	// this is so that we can keep pointers around to the blocks
 	// FIXME: if we cannot keep everything pinned in memory, we could switch to an out-of-memory merge join or so
 	for (auto &block : block_collection->blocks) {
 		auto handle = buffer_manager.Pin(block.block);
-		data_ptr_t dataptr = handle->node->buffer;
+		data_ptr_t entries_buffer = handle->node->buffer;
 		idx_t entry = 0;
 		while (entry < block.count) {
 			// fetch the next vector of entries from the blocks
-			idx_t next = MinValue<idx_t>(STANDARD_VECTOR_SIZE, block.count - entry);
-			for (idx_t i = 0; i < next; i++) {
-				hash_data[i] = Load<hash_t>((data_ptr_t)(dataptr + pointer_offset));
-				key_locations[i] = dataptr;
-				dataptr += entry_size;
+			idx_t entries_count = MinValue<idx_t>(STANDARD_VECTOR_SIZE, block.count - entry);
+			for (idx_t i = 0; i < entries_count; i++) {
+				hash_data[i] = Load<hash_t>((data_ptr_t)(entries_buffer + pointer_offset));
+				key_locations[i] = entries_buffer;
+				entries_buffer += entry_size; // walk to next tuple
 			}
-			// now insert into the hash table
-			InsertHashes(hashes, next, key_locations);
-
-			entry += next;
+			// now insert the key_locations into the hash table
+			InsertHashes(hashes, entries_count, key_locations);
+			entry += entries_count; // walk entries_count
 		}
 		pinned_handles.push_back(move(handle));
 	}
-
+	// In case of primary key with an inner join do a semi-join instead
+	if (join_type == JoinType::INNER && has_primary_key) {
+		join_type = JoinType::SEMI;
```

Comment:
Do we need to change the join type? It is not actually a semi join, it is still a regular inner join, just with an optimization applied.

Can't we add a function e.g. `bool IsUniqueInner()` that is used to select the faster implementation that does something like `join_type == JoinType::INNER && has_unique_keys`?

Reply:
Yeah, I agree it is best to have a different path in this case. I will implement that. Thanks.